content/realtime/authentication.textile

```diff
@@ -77,42 +77,110 @@ Here is a straightforward example of using Basic Authentication to connect to Ab
 ```
 
 ```[java]
-  ClientOptions options = new ClientOptions();
-  options.authUrl = "http://my.website/auth";
-  AblyRealtime realtime = new AblyRealtime(options);
+ClientOptions options = new ClientOptions();
+options.authUrl = "/auth";
+AblyRealtime realtime = new AblyRealtime(options);
 ```
 
 ```[objc]
-  ARTClientOptions *options = [[ARTClientOptions alloc] init];
-  options.authUrl = [NSURL URLWithString:@"http://my.website/auth"];
-  ARTRealtime *realtime = [[ARTRealtime alloc] initWithOptions:options];
+ARTClientOptions *options = [[ARTClientOptions alloc] init];
+options.authUrl = [NSURL URLWithString:@"/auth"];
+ARTRealtime *realtime = [[ARTRealtime alloc] initWithOptions:options];
 ```
 
 ```[swift]
-  let options = ARTClientOptions()
-  options.authUrl = NSURL(string: "http://my.website/auth")
-  let realtime = ARTRealtime(options: options)
+let options = ARTClientOptions()
+options.authUrl = NSURL(string: "/auth")
+let realtime = ARTRealtime(options: options)
 ```
 
 ```[csharp]
-  ClientOptions options = new ClientOptions();
-  options.AuthUrl = new Uri("http://my.website/auth");
-  AblyRealtime realtime = new AblyRealtime(options);
+ClientOptions options = new ClientOptions();
+options.AuthUrl = new Uri("/auth");
+AblyRealtime realtime = new AblyRealtime(options);
+```
+
+h4(#auth-options). How to specify auth options
+
+There are several "auth options":/realtime/authentication#auth-options you can specify along with @authUrl@ and @authCallback@:
+
+* @authMethod@ - when @authUrl@ is called, the default "GET" method will be used, unless "POST" is specified.
+* @authHeaders@ - allows you to pass additional headers as required, depending on your use case.
+* @authParams@ - allows you to pass additional query parameters, depending on your use case.
+
+```[jsall]
+const ably = new Ably.Realtime({ authUrl: "/auth", authMethod: "POST", authParams: { p1: param1, b: param2}, authHeaders: {h1: header1, h2: header2} });
+```
+
+h3(#auth-callback). Token refresh using @authCallback@
+
+For non-browser clients, an authentication callback is the preferred method of implementing token refresh, although browser clients can also use @authCallback@ if the use case requires it. This relies on your application communicating securely with your own servers to obtain a token.
+
+The following code sample shows an example of using an @authCallback@ in a client:
+
+```[jsall](code-editor:realtime/auth-token-callback)
+const https = require("https");
+require("dotenv").config();
+
+const options = {
+  hostname: process.env.HOSTNAME,
+  port: 443, // SSL
+  path: "/auth",
+  method: "POST",
+  headers: {
+    "User-Agent": "Realtime Client v0.2",
+    username: process.env.USERNAME,
+    password: process.env.PASSWORD,
+  },
+};
+
+const ably = new require("ably").Realtime({
+  authCallback: (tokenParams, callback) => {
+    const req = https.request(options, (res) => {
+      if (res.statusCode == 200) {
+        res.on("data", (tokenRequest) => {
```

Comment:
This usage of the NodeJS HTTP client is a little bit off; the "data" event can be emitted more than once if the response is sufficiently large (I'm not sure exactly what the conditions are for this to happen, I'd assume it depends on the server implementation). I think generally an Ably token request should be small enough to come through as a single chunk so maybe it's fine but it might raise a few eyebrows. The more 'correct' way of doing this is to initialise a string and append each chunk to the string whenever "data" is emitted. Then you can do the JSON parsing once "end" is emitted:
```
let tokenRequest = '';

res.on("data", (chunk) => {
  tokenRequest += chunk;
});

res.on("end", () => {
  callback(null, JSON.parse(tokenRequest));
});
```

Worth noting as well that JSON.parse can throw an error if the string supplied to it is not valid JSON. In a real world application I would definitely advise adding a try/catch block, although I realise we want these examples to be simple so I'm on the fence about adding one here.

Reply:
Yeah I did consider mentioning fetch earlier. The fetch API is obviously a lot simpler and I would assume the vast majority of people implementing auth with Ably won't want to use a low-level HTTP client (like the one in Node's standard lib). That being said, this particular snippet is supposedly aimed at Node users and fetch still isn't available in any LTS Node version so we'd at the very least need to recommend using a polyfill (such as node-fetch) for users who aren't using Node 17.5+. Definitely worth considering but I feel like it's gonna be noisy either way 🤷

Reply:
Whilst `fetch` isn't in LTS yet, I assume it'll be in the foreseeable future, then this code using `https` will seem outdated then?

Reply:
Yeah that's a good point. I've thought about it a bit more now and I'd be in favour of using fetch 👍

Reply:
Duly noted, fetch seems to be the way to go.